tools/outline_proxy_controller/network_monitor.cpp

```diff
@@ -0,0 +1,139 @@
+
+namespace outline {
+
+static_assert(std::is_integral<std::underlying_type_t<NetworkChangeEvent>>::value);
+
+NetworkChangeEvent operator|(NetworkChangeEvent lhs, NetworkChangeEvent rhs) noexcept {
+  using IntType = std::underlying_type_t<NetworkChangeEvent>;
+  return static_cast<NetworkChangeEvent>(static_cast<IntType>(lhs) | static_cast<IntType>(rhs));
+}
+
+NetworkChangeEvent& operator|=(NetworkChangeEvent& lhs, NetworkChangeEvent rhs) noexcept {
+  return lhs = lhs | rhs;
+}
+
+NetworkChangeEvent operator&(NetworkChangeEvent lhs, NetworkChangeEvent rhs) noexcept {
+  using IntType = std::underlying_type_t<NetworkChangeEvent>;
+  return static_cast<NetworkChangeEvent>(static_cast<IntType>(lhs) & static_cast<IntType>(rhs));
+}
+
+std::string to_string(NetworkChangeEvent evt) {
+  if (evt == NetworkChangeEvent::kNone) {
+    return "None";
+  }
+  std::string result{};
+  if ((evt & NetworkChangeEvent::kNicChanged) == NetworkChangeEvent::kNicChanged) {
+    result += "NIC";
+  }
+  if ((evt & NetworkChangeEvent::kAddressChanged) == NetworkChangeEvent::kAddressChanged) {
+    result += (result.empty() ? "IP" : " IP");
+  }
+  if ((evt & NetworkChangeEvent::kRouteChanged) == NetworkChangeEvent::kRouteChanged) {
+    result += (result.empty() ? "Route" : " Route");
+  }
+  return result;
+}
+
+}
+
+NetworkMonitor::NetworkMonitor(const boost::asio::any_io_executor &io_context)
+  : netlink_socket_{io_context}
+{
+  using namespace boost::asio::generic;
+
+  sockaddr_nl sa;
+  std::memset(&sa, 0, sizeof(sa));
+  sa.nl_family = AF_NETLINK;
+  sa.nl_groups = RTMGRP_LINK         // network card change
+               | RTMGRP_IPV4_IFADDR  // IPv4 address change
+               | RTMGRP_IPV4_ROUTE   // IPv4 route table change
+               | RTMGRP_IPV6_IFADDR  // IPv6 address change
+               | RTMGRP_IPV6_ROUTE;  // IPv6 route table change
+
+  netlink_socket_.open({AF_NETLINK, NETLINK_ROUTE});
+  netlink_socket_.bind({&sa, sizeof(sa)});
+
+  logger.info("network monitor initialized");
+}
+
+NetworkMonitor::~NetworkMonitor() {
+  logger.info("network monitor destroyed");
+}
+
+boost::asio::awaitable<NetworkChangeEvent> NetworkMonitor::WaitForChangeEvent() {
+  using namespace boost::asio;
+
+  std::array<std::byte, kRecvBufferSize> buf;
+  auto received_events = NetworkChangeEvent::kNone;
+
+  do {
+    auto len = co_await netlink_socket_.async_receive(boost::asio::buffer(buf), use_awaitable);
+
+    // https://linux.die.net/man/7/netlink
+    for (auto msg = reinterpret_cast<nlmsghdr*>(buf.data()); NLMSG_OK(msg, len); msg = NLMSG_NEXT(msg, len)) {
+      switch (msg->nlmsg_type) {
+        case NLMSG_DONE:
+          goto EndOfMultiPartMsg;
+        case NLMSG_ERROR:
+        {
+          auto err = reinterpret_cast<nlmsgerr*>(NLMSG_DATA(msg));
+          throw std::system_error{err->error, std::system_category()};
+        }
+        case RTM_NEWLINK:
+          received_events |= NetworkChangeEvent::kNicChanged;
```

Comment:
This is subtle and error prone. Let's define a function or type method instead instead. Things like `received_events.Set(kNicChanged)` or `SetEvent(received_events, kNicChanged)` are clearer.

See https://google.github.io/styleguide/cppguide.html#Operator_Overloading

Reply:
Also, we shouldn't expose bit operations to the user of the library. That's an implementation detail the user shouldn't have to care about.

Reply:
I was looking for something similar to Java's [`EnumSet<>`](https://docs.oracle.com/javase/8/docs/api/java/util/EnumSet.html) and found nothing in C++, so I decided to imitate the old bit-flag integer, but that requires the caller know the difference about `==`  and `( & ) ==`. It makes sense to use methods such as `HasRouteChanged()`, `HasNicChanged()` to hide the implementation detail.

Reply:
I didn't get, did you mean you were going to add those methods to hide the implementation? I don't see it.

Reply:
code updated, basically i exposed the `NetworkChangeEvent` as a `class` instead of an bit-wise `enum` flag.